x/virtualgroup/keeper/keeper.go

```diff
@@ -647,3 +647,283 @@ func (k Keeper) CompleteSwapOut(ctx sdk.Context, gvgFamilyID uint32, gvgIDs []ui
+			return sptypes.ErrStorageProviderNotFound.Wrapf("spID(%d) is different from the spID(%d) in swapInInfo", successorSPID, swapInInfo.SuccessorSpId)
+		}
+		store.Delete(key)
+	}
+
+	if err := ctx.EventManager().EmitTypedEvents(&types.EventCancelSwapIn{
+		StorageProviderId:          swapInInfo.SuccessorSpId,
+		GlobalVirtualGroupFamilyId: gvgFamilyID,
+		GlobalVirtualGroupId:       gvgID,
+		TargetSpId:                 swapInInfo.TargetSpId,
+	}); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (k Keeper) CompleteSwapIn(ctx sdk.Context, gvgFamilyID uint32, gvgID uint32, successorSP *sptypes.StorageProvider) error {
+	store := ctx.KVStore(k.storeKey)
+	swapInInfo := types.SwapInInfo{}
+	// swapIn family
+	if gvgFamilyID != types.NoSpecifiedFamilyId {
+		key := types.GetSwapInFamilyKey(gvgFamilyID)
+		bz := store.Get(key)
+		if bz == nil {
+			return types.ErrSwapInFailed.Wrapf("The swap info not found in blockchain.")
+		}
+		k.cdc.MustUnmarshal(bz, &swapInInfo)
+		if successorSP.Id != swapInInfo.SuccessorSpId {
+			return types.ErrSwapInFailed.Wrapf("The SP(ID: %d) has not reserved the swap(swapInfo=%s)", successorSP.Id, swapInInfo.String())
+		}
+		if uint64(ctx.BlockTime().Unix()) >= swapInInfo.ExpirationTime {
+			return types.ErrSwapInFailed.Wrapf("reserved swap expired")
+		}
+
+		targetPrimarySP, found := k.spKeeper.GetStorageProvider(ctx, swapInInfo.TargetSpId)
+		if !found {
+			return sptypes.ErrStorageProviderNotFound.Wrapf("The storage provider(ID: %d) not found when complete swap in.", swapInInfo.TargetSpId)
+		}
+		if err := k.completeSwapInFamily(ctx, successorSP, targetPrimarySP, gvgFamilyID); err != nil {
+			return err
+		}
+		store.Delete(key)
+	} else {
+		key := types.GetSwapInGVGKey(gvgID)
+		bz := store.Get(key)
+		if bz == nil {
+			return types.ErrSwapInFailed.Wrapf("The swap info not found in blockchain.")
+		}
+		k.cdc.MustUnmarshal(bz, &swapInInfo)
+		if successorSP.Id != swapInInfo.SuccessorSpId {
+			return types.ErrSwapInFailed.Wrapf("The sp(ID: %d) has not reserved the swap for secondary SP(ID: %d)", successorSP.Id, swapInInfo.TargetSpId)
+		}
+		targetSecondarySP, found := k.spKeeper.GetStorageProvider(ctx, swapInInfo.TargetSpId)
+		if !found {
+			return sptypes.ErrStorageProviderNotFound.Wrapf("The storage provider(ID: %d) not found when complete swap in.", swapInInfo.TargetSpId)
+		}
+		if err := k.completeSwapInGVG(ctx, successorSP.Id, targetSecondarySP.Id, gvgID); err != nil {
+			return err
+		}
+		store.Delete(key)
+	}
+	if err := ctx.EventManager().EmitTypedEvents(&types.EventCompleteSwapIn{
+		StorageProviderId:          successorSP.Id,
+		TargetStorageProviderId:    swapInInfo.TargetSpId,
+		GlobalVirtualGroupFamilyId: gvgFamilyID,
+		GlobalVirtualGroupId:       gvgID,
+	}); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (k Keeper) completeSwapInFamily(ctx sdk.Context, successorSP, targetPrimarySP *sptypes.StorageProvider, familyID uint32) error {
```

Comment:
Does this code copy from the complete swap out?

If yes, can we reuse the code

Reply:
Are we going to remove all swapOut related msg or keep them?

Reply:
adjusted